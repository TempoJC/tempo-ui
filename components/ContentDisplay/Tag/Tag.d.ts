import { TagProps } from './types';
/**
 * Component that shows a text value as a tag.
 */
declare const Tag: ({ className, onClick, onRemoveButtonClick, readOnly, label, disabled, color, variant, testId, uppercase, }: TagProps) => import("react/jsx-runtime").JSX.Element;
export default Tag;
