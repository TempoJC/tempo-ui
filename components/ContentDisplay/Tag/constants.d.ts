export declare const TAG_DARK = "dark";
export declare const TAG_LIGHT = "light";
export declare const TAG_COLOR_COCONUT = "coconut";
export declare const TAG_COLOR_BLACKBERRY = "blackberry";
export declare const TAG_COLOR_CHERRY = "cherry";
export declare const TAG_COLOR_APRICOT = "apricot";
export declare const TAG_COLOR_KIWI = "kiwi";
export declare const TAG_COLOR_MINT = "mint";
export declare const TAG_COLOR_WATER = "water";
export declare const TAG_COLOR_BLUEBERRY = "blueberry";
export declare const TAG_COLOR_GRAPE = "grape";
export declare const TAG_COLOR_WATERMELON = "watermelon";
