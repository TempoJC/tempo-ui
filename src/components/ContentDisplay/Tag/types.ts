import { CommonProps, VariantColor } from "@/types";
import type { MouseEventHandler, ReactNode } from "react";

export type TagProps = CommonProps & {
  /**
   * Set to `true` to show the component disabled.
   */
  disabled?: boolean;
  /**
   * Tag label text.
   */
  label: ReactNode;
  /**
   * Callback function fired when tag is clicked.
   */
  onClick?: MouseEventHandler<HTMLDivElement>;
  /**
   * Function that will be set as the `onClick` handler of the deletion button shown
   * when this prop is defined and `readOnly` is false.
   */
  onRemoveButtonClick?: MouseEventHandler<HTMLSpanElement>;
  /**
   * Indicates whether the tag is read-only.
   *
   * Use `false` to show the tag deletion button and do not forget to define the
   * `onRemoveButtonClick` handler.
   */
  readOnly?: boolean;
  /**
   * Set to `false` to show the original label instead of capital letters.
   */
  uppercase?: boolean;
  /**
   * Color variant.
   */
  variant?: VariantColor;
};
