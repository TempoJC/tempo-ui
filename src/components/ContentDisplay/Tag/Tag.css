.tempo-tag--light {
  --tempo-tag-color: var(--color-token-light-content-1);
  --tempo-tag-color-hover: var(--color-token-light-content-hover);
  --tempo-tag-color-active: var(--color-token-light-content-1);
  --tempo-tag-color-disabled: var(--color-token-light-content-disabled);
  --tempo-tag-background: var(--color-token-light-background-2);
  --tempo-tag-background-hover: var(--color-token-light-background-hover);
  --tempo-tag-background-active: var(--color-token-light-background-active);
  --tempo-tag-background-disabled: var(--color-token-light-background-disabled);
}

.tempo-tag--dark {
  --tempo-tag-color: var(--color-token-dark-content-1);
  --tempo-tag-color-hover: var(--color-token-dark-content-hover);
  --tempo-tag-color-active: var(--color-token-dark-content-1);
  --tempo-tag-color-disabled: var(--color-token-dark-content-disabled);
  --tempo-tag-background: var(--color-token-dark-background-2);
  --tempo-tag-background-hover: var(--color-token-dark-background-hover);
  --tempo-tag-background-active: var(--color-token-dark-background-active);
  --tempo-tag-background-disabled: var(--color-token-dark-background-disabled);
}

.tempo-tag--color {
  --tempo-tag-color: var(--color-token-light-content-1);
  --tempo-tag-color-hover: var(--tempo-tag-color-hover);
  --tempo-tag-color-active: var(--color-token-light-content-inverse);
}

.tempo-tag--color-coconut {
  --tempo-tag-background: var(--color-coconut200);
  --tempo-tag-background-hover: var(--color-coconut300);
  --tempo-tag-background-active: var(--color-coconut400);
}

.tempo-tag--color-blackberry {
  --tempo-tag-background: var(--color-blackberry200);
  --tempo-tag-background-hover: var(--color-blackberry300);
  --tempo-tag-background-active: var(--color-blackberry400);
}

.tempo-tag--color-cherry {
  --tempo-tag-background: var(--color-cherry200);
  --tempo-tag-background-hover: var(--color-cherry300);
  --tempo-tag-background-active: var(--color-cherry400);
}

.tempo-tag--color-apricot {
  --tempo-tag-background: var(--color-apricot200);
  --tempo-tag-background-hover: var(--color-apricot300);
  --tempo-tag-background-active: var(--color-apricot400);
}

.tempo-tag--color-kiwi {
  --tempo-tag-background: var(--color-kiwi200);
  --tempo-tag-background-hover: var(--color-kiwi300);
  --tempo-tag-background-active: var(--color-kiwi400);
}

.tempo-tag--color-mint {
  --tempo-tag-background: var(--color-mint200);
  --tempo-tag-background-hover: var(--color-mint300);
  --tempo-tag-background-active: var(--color-mint400);
}

.tempo-tag--color-water {
  --tempo-tag-background: var(--color-water200);
  --tempo-tag-background-hover: var(--color-water300);
  --tempo-tag-background-active: var(--color-water400);
}

.tempo-tag--color-blueberry {
  --tempo-tag-background: var(--color-blueberry200);
  --tempo-tag-background-hover: var(--color-blueberry300);
  --tempo-tag-background-active: var(--color-blueberry400);
}

.tempo-tag--color-grape {
  --tempo-tag-background: var(--color-grape200);
  --tempo-tag-background-hover: var(--color-grape300);
  --tempo-tag-background-active: var(--color-grape400);
}

.tempo-tag--color-watermelon {
  --tempo-tag-background: var(--color-watermelon200);
  --tempo-tag-background-hover: var(--color-watermelon300);
  --tempo-tag-background-active: var(--color-watermelon400);
}

.tempo-tag {
  white-space: nowrap;
  display: inline-flex;
  flex-direction: row;
  align-items: center;
  padding: var(--size1) var(--size2);
  gap: 2px;
  background: var(--tempo-tag-background);
}

.tempo-tag:has(.tempo-tag__action) {
  padding: 3px 2px 3px var(--size2);
}

.tempo-tag__label {
  font: var(--typography-label-l-regular-font);
  letter-spacing: var(--typography-label-l-regular-letter-spacing);
  text-transform: var(--typography-label-l-regular-text-transform);
  user-select: none;
  color: var(--tempo-tag-color);
}

.tempo-tag__label--lowercase {
  text-transform: none;
}

.tempo-tag:hover {
  background: var(--tempo-tag-background-hover);
}

.tempo-tag:has(.tempo-tag__action:hover) {
  background: var(--tempo-tag-background);
}

.tempo-tag:has(.tempo-tag__action:hover) .tempo-tag__label {
  color: var(--tempo-tag-color);
}

.tempo-tag:hover .tempo-tag__label {
  color: var(--tempo-tag-color-hover);
}

.tempo-tag:active {
  background: var(--tempo-tag-background-active);
}

.tempo-tag:active .tempo-tag__label {
  color: var(--tempo-tag-color-active);
}

.tempo-tag--disabled {
  pointer-events: none;
  background: var(--tempo-tag-background-disabled);
}

.tempo-tag.tempo-tag--disabled .tempo-tag__label {
  color: var(--tempo-tag-color-disabled);
}

.tempo-tag__action {
  width: var(--size5);
  height: var(--size5);
  display: block;
  cursor: pointer;
}

.tempo-tag__action path {
  fill: var(--tempo-tag-color);
}

.tempo-tag--disabled .tempo-tag__action,
.tempo-tag--disabled .tempo-tag__action:hover {
  pointer-events: none;
  background: none;
}

.tempo-tag--disabled .tempo-tag__action path,
.tempo-tag--disabled .tempo-tag__action:hover path {
  fill: var(--tempo-tag-color-disabled);
  pointer-events: none;
}

.tempo-tag__action:hover {
  background: var(--tempo-tag-background-hover);
}

.tempo-tag:hover path,
.tempo-tag__action:hover path {
  fill: var(--tempo-tag-color-hover);
}

.tempo-tag__action:active {
  background: var(--tempo-tag-background-active);
}

.tempo-tag:active .tempo-tag__action path,
.tempo-tag__action:active path {
  fill: var(--tempo-tag-color-active);
}
