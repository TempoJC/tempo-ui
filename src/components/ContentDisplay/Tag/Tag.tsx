import React from "react";
import cn from "classnames";
import { Icon } from "@/components";
import { TagProps } from "./types";
import {
  TAG_COLOR_APRICOT,
  TAG_COLOR_BLACKBERRY,
  TAG_COLOR_BLUEBERRY,
  TAG_COLOR_CHERRY,
  TAG_COLOR_COCONUT,
  TAG_COLOR_GRAPE,
  TAG_COLOR_KIWI,
  TAG_COLOR_MINT,
  TAG_COLOR_WATER,
  TAG_COLOR_WATERMELON,
  TAG_DARK,
  TAG_LIGHT,
} from "./constants";
import "./Tag.css";

/**
 * Component that shows a text value as a tag.
 */
const Tag = ({
  className,
  onClick,
  onRemoveButtonClick,
  readOnly = true,
  label,
  disabled = false,
  color = TAG_LIGHT,
  variant,
  testId,
  uppercase = true,
}: TagProps) => {
  const handleRemove = (e: React.MouseEvent<HTMLSpanElement> | React.KeyboardEvent<HTMLSpanElement>) => {
    if (disabled) return undefined;
    e.stopPropagation();

    onRemoveButtonClick?.(e as React.MouseEvent<HTMLSpanElement>);
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLSpanElement>) => {
    if (e.key === "Enter" || e.key === " ") {
      e.preventDefault();
      handleRemove(e);
    }
  };

  return (
    <div
      className={cn(
        "tempo-tag",
        className,
        {
          "tempo-tag--dark": color === TAG_DARK,
          "tempo-tag--light": color === TAG_LIGHT,
        },
        {
          "tempo-tag--color": Boolean(variant),
          "tempo-tag--color-coconut": variant === TAG_COLOR_COCONUT,
          "tempo-tag--color-blackberry": variant === TAG_COLOR_BLACKBERRY,
          "tempo-tag--color-cherry": variant === TAG_COLOR_CHERRY,
          "tempo-tag--color-apricot": variant === TAG_COLOR_APRICOT,
          "tempo-tag--color-kiwi": variant === TAG_COLOR_KIWI,
          "tempo-tag--color-mint": variant === TAG_COLOR_MINT,
          "tempo-tag--color-water": variant === TAG_COLOR_WATER,
          "tempo-tag--color-blueberry": variant === TAG_COLOR_BLUEBERRY,
          "tempo-tag--color-grape": variant === TAG_COLOR_GRAPE,
          "tempo-tag--color-watermelon": variant === TAG_COLOR_WATERMELON,
        },
        {
          "tempo-tag--disabled": disabled,
          "tempo-tag--readOnly": readOnly,
        },
      )}
      data-testid={testId}
    >
      <div
        onClick={disabled ? undefined : onClick}
        className={cn("tempo-tag__label", {
          "tempo-tag__label--lowercase": !uppercase,
        })}
        data-testid={testId && `${testId}-label`}
      >
        {label}
      </div>
      {!readOnly && (
        <span
          className="tempo-tag__action"
          onClick={handleRemove}
          onKeyDown={handleKeyDown}
          role="button"
          tabIndex={disabled ? -1 : 0}
          data-testid={testId && `${testId}-delete-button`}
        >
          <Icon group="actions" name="close-small" />
        </span>
      )}
    </div>
  );
};

export default Tag;
