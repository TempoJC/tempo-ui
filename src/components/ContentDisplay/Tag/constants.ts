export const TAG_DARK = "dark";
export const TAG_LIGHT = "light";

// Tag variant colors
export const TAG_COLOR_COCONUT = "coconut";
export const TAG_COLOR_BLACKBERRY = "blackberry";
export const TAG_COLOR_CHERRY = "cherry";
export const TAG_COLOR_APRICOT = "apricot";
export const TAG_COLOR_KIWI = "kiwi";
export const TAG_COLOR_MINT = "mint";
export const TAG_COLOR_WATER = "water";
export const TAG_COLOR_BLUEBERRY = "blueberry";
export const TAG_COLOR_GRAPE = "grape";
export const TAG_COLOR_WATERMELON = "watermelon";
