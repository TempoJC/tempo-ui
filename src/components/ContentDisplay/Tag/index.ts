// packages/components-content/tag
export { default as Tag } from "./Tag";
export * from "./types";
