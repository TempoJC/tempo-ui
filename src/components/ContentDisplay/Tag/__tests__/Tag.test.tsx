import { userEvent } from "@testing-library/user-event";
import { render, screen } from "@testing-library/react";
import Tag from "../Tag";

describe("@tempo-ui/ContentDisplay/Tag", () => {
  it("renders", () => {
    const deleteHandler = () => {};
    const { asFragment } = render(<Tag label="label" onRemoveButtonClick={deleteHandler} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("calls the tag click handler", async () => {
    const clickHandler = jest.fn();

    render(<Tag onClick={clickHandler} label="label" />);

    await userEvent.click(screen.getByText("label"));

    expect(clickHandler).toHaveBeenCalledTimes(1);
  });

  it("renders with other variant", () => {
    render(<Tag label="label" variant="blueberry" />);

    expect(screen.getByText("label").parentElement).toHaveClass("tempo-tag--color-blueberry");
  });

  describe("when deletable", () => {
    it("renders a delete button", () => {
      render(<Tag label="label" readOnly={false} />);

      expect(screen.getByRole("button")).toBeInTheDocument();
      expect(screen.getByRole("button")).toHaveClass("tempo-tag__action");
    });

    it("calls the delete handler", async () => {
      const clickHandler = jest.fn();
      const deleteHandler = jest.fn();

      render(<Tag onRemoveButtonClick={deleteHandler} onClick={clickHandler} label="label" readOnly={false} />);

      await userEvent.click(screen.getByRole("button"));

      expect(deleteHandler).toHaveBeenCalledTimes(1);
      expect(clickHandler).not.toHaveBeenCalled();
    });
  });

  it("adds the additional className", () => {
    render(<Tag className="customClass" label="label" readOnly />);

    expect(screen.getByText("label").parentElement).toHaveClass("customClass");
  });

  describe("when disabled", () => {
    it("renders a disabled tag", () => {
      render(<Tag label="label" disabled />);

      expect(screen.getByText("label").parentElement).toHaveClass("tempo-tag--disabled");
    });

    it("does not call the click handler", async () => {
      const clickHandler = jest.fn();

      render(<Tag onClick={clickHandler} label="label" disabled />);

      await userEvent.click(screen.getByText("label"));

      expect(clickHandler).not.toHaveBeenCalled();
    });

    it("does not call the delete handler", async () => {
      const deleteHandler = jest.fn();

      render(<Tag onRemoveButtonClick={deleteHandler} label="label" disabled readOnly={false} />);

      await userEvent.click(screen.getByRole("button"));

      expect(deleteHandler).not.toHaveBeenCalled();
    });
  });

  describe("testId prop", () => {
    it("renders the Tag with the specified data-testid attribute", () => {
      const { getByTestId } = render(<Tag label="label" testId="tag" />);
      const element = getByTestId("tag");
      expect(element).toBeInTheDocument();
      expect(element).toHaveAttribute("data-testid", "tag");
    });
  });

  it("render lowercase label", () => {
    render(<Tag label="label" testId="tag" uppercase={false} />);
    expect(screen.getByText("label")).toHaveClass("tempo-tag__label--lowercase");
  });
});
