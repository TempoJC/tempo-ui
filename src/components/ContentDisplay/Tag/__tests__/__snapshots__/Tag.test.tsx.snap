// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/ContentDisplay/Tag renders 1`] = `
<DocumentFragment>
  <div
    class="tempo-tag tempo-tag--light tempo-tag--readOnly"
  >
    <div
      class="tempo-tag__label"
    >
      label
    </div>
  </div>
</DocumentFragment>
`;
